src/extra/recipes/vs_utils.py

```diff
@@ -0,0 +1,245 @@
+        explained_variance = S**2 / (self.n_samples_seen_ - 1)
+        explained_variance_ratio = S**2 / np.sum(col_var * n_total_samples)
+
+        self.components_ = Vt[: self.n_components_]
+        self.singular_values_ = S[: self.n_components_]
+        self.mean_ = col_mean
+        self.var_ = col_var
+        self.explained_variance_ = explained_variance[: self.n_components_]
+        self.explained_variance_ratio_ = explained_variance_ratio[: self.n_components_]
+        # we already checked `self.n_components <= n_samples` above
+        if self.n_components_ not in (n_samples, n_features):
+            self.noise_variance_ = explained_variance[self.n_components_ :].mean()
+        else:
+            self.noise_variance_ = 0.0
+        #print(f'rest = {time.time()-start}s')
+        return self
+
+def _fit_estimator(estimator, X: np.ndarray, y: np.ndarray, w: Optional[np.ndarray]=None):
+    from sklearn.base import clone
+    estimator = clone(estimator)
+    if w is None:
+        estimator.fit(X, y)
+    else:
+        estimator.fit(X, y, w)
+    return estimator
+
+def get_properties(model):
+    return {k: v for k, v in model.__dict__.items() if k.endswith('_')}
+
+class MultiOutputGenericWithStd(MetaEstimatorMixin, BaseEstimator):
+
+    def __init__(self, estimator, *, n_jobs=-1):
+        self.estimator = estimator
+        self.n_jobs = n_jobs
+
+    def fit(self, X: np.ndarray, y: np.ndarray, weights: Optional[np.ndarray]=None):
+        """Fit the model to data, separately for each output variable.
+
+        Args:
+          X: {array-like, sparse matrix} of shape (n_samples, n_features)
+            The input data.
+          y: {array-like, sparse matrix} of shape (n_samples, n_outputs)
+            Multi-output targets. An indicator matrix turns on multilabel
+            estimation.
+
+        Returns: self.
+        """
+        from joblib import Parallel, delayed
+        if y.ndim == 1:
+            raise ValueError(
+                "y must have at least two dimensions for "
+                "multi-output regression but has only one."
+            )
+        self.estimators_ = Parallel(n_jobs=self.n_jobs)(
+            delayed(_fit_estimator)(
+                self.estimator, X, y[:, i], weights
+            )
+            for i in range(y.shape[1])
+        )
+
+        return self
+
+    def predict(self, X: np.ndarray, return_std: bool=False) -> Union[np.ndarray, Tuple[np.ndarray, np.ndarray]]:
+        """Predict multi-output variable using model for each target variable.
+
+        Args:
+          X: {array-like, sparse matrix} of shape (n_samples, n_features)
+            The input data.
+
+        Returns: {array-like, sparse matrix} of shape (n_samples, n_outputs)
+            Multi-output targets predicted across multiple predictors.
+            Note: Separate models are generated for each predictor.
+        """
+        from joblib import Parallel, delayed
```

Comment:
`joblib` needs to be added to `pyproject.toml` if we're going to use it. But why use joblib over e.g. `multiprocessing.Pool.map()`?

Reply:
I agree. Done.

Reply:
I'm a bit confused, it looks like joblib is still being used?

Reply:
Ah now I see, it was replaced in `fit()` but not `predict()`. Is it feasible to remove from `predict()` too?

Reply:
I'm just a twat, who did not notice it is also used elsewhere. I pushed the update.